src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp

```diff
@@ -0,0 +1,239 @@
+
+/// \file
+/// Havoc multiple and possibly dependent targets simultaneously
+
+#include "havoc_assigns_clause_targets.h"
+
+#include <map>
+
+#include <langapi/language_util.h>
+
+#include <util/c_types.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
+#include <util/pointer_expr.h>
+#include <util/pointer_offset_size.h>
+#include <util/pointer_predicates.h>
+#include <util/std_code.h>
+
+#include "utils.h"
+
+static const char ASSIGNS_CLAUSE_REPLACEMENT_TRACKING[] =
+  " assigned by replaced ";
+
+/// Returns a pointer expression to the start address of the target
+///
+/// - If the target is a `pointer_object(p)` expression,
+///   return a pointer to the same object with offset zero
+/// - else, if the target is a sized and assignable expression,
+///   return its address
+/// - else trigger an error.
+///
+static exprt base_pointer(
+  const exprt &target,
+  // context parameters
+  const namespacet &ns)
+{
+  if(target.id() == ID_pointer_object)
+  {
+    const auto &ptr = target.operands().front();
+    // bring the offset back to zero
+    auto res = minus_exprt{ptr, pointer_offset(ptr)};
+    res.type() = ptr.type();
+    return std::move(res);
+  }
+  else if(is_assignable(target))
+  {
+    // check that the type of the target has a definite size
+    const auto &size = size_of_expr(target.type(), ns);
+
+    INVARIANT(
+      size.has_value(),
+      "`sizeof` must always be computable on assignable assigns clause "
+      "targets.");
+
+    return address_of_exprt{target, pointer_type(target.type())};
+  }
+  UNREACHABLE;
+}
+
+/// \brief Generates instructions to conditionally snapshot the value
+/// of `target_pointer` into `target_snapshot_var`.
+///
+/// ```
+/// DECL target_valid_var : bool
+/// DECL target_snapshot_var : <target_pointer.type()>
+/// ASSIGN target_valid_var := <all_dereferences_valid(target_pointer)>
+/// ASSIGN target_snapshot_var := NULL
+/// IF !target_valid_var GOTO skip_target
+/// ASSIGN target_snapshot_var := target_pointer;
+/// skip_target: SKIP
+/// ```
+///
+static void snapshot_if_valid(
+  const symbol_exprt &target_valid_var,
+  const symbol_exprt &target_snapshot_var,
+  const exprt &target_pointer,
+  goto_programt &dest,
+  // context parameters
+  const namespacet &ns)
+{
+  dest.add(goto_programt::make_decl(target_valid_var));
+
+  dest.add(goto_programt::make_decl(target_snapshot_var));
+
+  dest.add(goto_programt::make_assignment(
+    target_valid_var, all_dereferences_are_valid(target_pointer, ns)));
+
+  dest.add(goto_programt::make_assignment(
+    target_snapshot_var,
+    null_pointer_exprt{to_pointer_type(target_pointer.type())}));
+
+  goto_programt skip_program;
+  const auto skip_target = skip_program.add(goto_programt::make_skip());
+
+  dest.add(goto_programt::make_goto(skip_target, not_exprt{target_valid_var}));
+
+  dest.add(goto_programt::make_assignment(target_snapshot_var, target_pointer));
+
+  dest.destructive_append(skip_program);
```

Comment:
You could use `insert_before_swap` to avoid the `skip_program`. But then it probably doesn't make a whole lot of a difference.

Reply:
it seems less natural, since after inserting the skip statement in dest, you would have to use a mix of insert_before_swap and skip_target++ or to insert instructions in reverse, or to create another program and insert_before_swap and then ++ the skip_target

Reply:
Ok!